File: Stalin Sort/stalinsort.js
// Stalin sort is an "algorithm" that will look at a list and eliminate all the un-sorted numbers
// It returns the sorted, probably shorter, list
// Its advantage: it completes in linear time: O(n).
// Its disadvantage: it doesn't "really" sort as much as it eliminates the unsorted

function stalinSort(list) {
  if (!Array.isArray(list)) {
    throw new TypeError('You can only sort an Array!');
  }

  // We create a variable that will hold our sorted item
  // This guy is always sorted, always in the right order, everything starts with him
  let pivot;

  // We now filter the list and start looking at the all the items
  // by comparing each to the pivot item
  // We also update the pivot item every time we get to a new item that is in the right order
  return list.filter((item, idx) => {
    // At index 0, we have our pivot item, so we keep it in the list
    // If the current item is greater than or equal to the pivot item,
    // we update the pivot item for next time and we add the item to the list
    if (idx === 0 || item >= pivot) {
      pivot = item;
      return true;
    }

    // Anthing else is unsorted and should be discarded
    return false;
  });
}

// And now we can use it
// console.log(stalinSort([1,2,3,6,4,7]));

File: sort/Stalin Sort/stalin_sort.cpp
#include <iostream>
#include <vector>

void stalinSort(std::vector<int> &arr, std::vector<int> &sorted)
{
    sorted.push_back(arr[0]);
    for (int i = 1; i < arr.size(); i++)
    {
        if (arr[i] > sorted[sorted.size() - 1])
        {
            sorted.push_back(arr[i]);
        }
    }
}

int main()
{
    std::vector<int> arr = {1,4,6,8,3,5,7,9,20,18,29};
    std::vector<int> sortedArr;

    stalinSort(arr, sortedArr);

    std::cout << "Before Stalin Sort\n";
    for (int a : arr)
    {
        std::cout << a << " ";
    }
    std::cout << "\nAfter Stalin Sort: Unsorted numbers are sent to gulag!\n";
    for (int a : sortedArr)
    {
        std::cout << a << " ";
    }
    std::cout << "\n";

    return 0;
}